### PR TITLE
NVRTC kernels for cast-transpose

### PR DESCRIPTION
**Pull request by timmoon10**
This is a follow-up to https://github.com/NVIDIA/TransformerEngine/pull/138 that adds NVRTC kernels for cast-transpose.

I've also tweaked the kernel selection heuristics to be a bit more general. There are some performance differences at intermediate scale, but nothing too drastic:
<details> <summary> Transpose kernel times </summary>

image

Evaluated FP8 transposes on an H100 PCIe with square matrices with nice dims (power of two or halfway between powers of two).

</details>

### CODE REVIEW

**Comment by timmoon10** [comment]
If there are no objections, I will merge this after the 1.6 release.

File: transformer_engine/common/transpose/cast_transpose.cu [hunk 1]
```diff
@@ -6,432 +6,360 @@
...
+               size_t num_rows,
+               size_t itype_size,
+               size_t otype_size,
+               size_t load_size_,
+               size_t store_size_)
+    : load_size{load_size_}
+    , store_size{store_size_} {
+    // Check that tiles are correctly aligned
+    constexpr size_t cache_line_size = 128;
+    if (load_size % itype_size != 0
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/transpose/cast_transpose.cu:62]: Maybe to add a notice that the object has not been fully constructed and might be in inconsistent state?

File: transformer_engine/common/transpose/cast_transpose.cu [hunk 2]
```diff
@@ -6,432 +6,360 @@
...
-    cast_and_transpose_regs<true>(in[current_in ^ 1], out_trans, my_output_c_tile,
-                                  current_place, stride, max, scale, true);
-#pragma unroll
-    for (unsigned int j = 0; j < nvec_in; ++j) {
-      out_space[i][j].data.vec = out_trans[j].data.vec;
+    const size_t row_tile_elements = load_size * THREADS_PER_WARP / itype_size;
+    const size_t col_tile_elements = store_size * THREADS_PER_WARP / otype_size;
+    valid = (row_length % row_tile_elements == 0
+             && num_rows % col_tile_elements == 0);
+    if (!valid) {
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/transpose/cast_transpose.cu:72]: Maybe to add a notice that the object has not been fully constructed and might be in inconsistent state?

File: transformer_engine/common/transpose/cast_transpose.cu [hunk 3]
```diff
@@ -6,432 +6,360 @@
...
-                                 OType * const output_c,
-                                 OType * const output_t,
-                                 const CType * const scale_ptr,
-                                 CType * const amax,
-                                 const size_t row_length,
-                                 const size_t num_rows,
-                                 const size_t num_tiles) {
+__launch_bounds__(block_size)
+cast_transpose_general_kernel(const IType * __restrict__ const input,
+                              const CType * const noop,
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/transpose/cast_transpose.cu:120]: Was the __restrict__ keyword intentionally omitted here?

File: transformer_engine/common/transpose/transpose.cu [hunk 4]
```diff
@@ -25,7 +27,80 @@ namespace {
...
+  KernelConfig(size_t row_length,
+               size_t num_rows,
+               size_t type_size,
+               size_t load_size_,
+               size_t store_size_)
+    : load_size{load_size_}
+    , store_size{store_size_} {
+    // Check that tiles are correctly aligned
+    constexpr size_t cache_line_size = 128;
+    if (load_size % type_size != 0
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/transpose/transpose.cu:58]: Maybe to add a notice that the object has not been fully constructed and might be in inconsistent state?

File: transformer_engine/common/transpose/transpose.cu [hunk 5]
```diff
@@ -25,7 +27,80 @@ namespace {
...
+        || cache_line_size % type_size != 0) {
+      return;
+    }
+    const size_t row_tile_elements = load_size * THREADS_PER_WARP / type_size;
+    const size_t col_tile_elements = store_size * THREADS_PER_WARP / type_size;
+    valid = (row_length % row_tile_elements == 0
+             && num_rows % col_tile_elements == 0);
+    if (!valid) {
+      return;
+    }
```
Comment (Oleg-Goncharov) [inline transformer_engine/common/transpose/transpose.cu:69]: Maybe to add a notice that the object has not been fully constructed and might be in inconsistent state?

**Review by Oleg-Goncharov — APPROVED** [review]
LGTM, besides the early return from the kernel configuration constructors

**Pull request merged**